Дипломный проект. Основное задание

Kirill495 opened the pull request.

Commit:
Спринт 18.
Дипломный проект. Разработка основного сервиса.

Commit:
Спринт 18.
Дипломный проект. Разработка основного сервиса. Добавлены подборки

Commit:
Спринт 18.
Дипломный проект. Разработка основного сервиса.

Commit:
Спринт 18.
Дипломный проект. Разработка основного сервиса. Добавлена проверка на даты в сервис статистики.

Commit:
Спринт 18.
Дипломный проект. Разработка основного сервиса. Доработка по результатам самостоятельного code-review.

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/api_admin/AdmUserController.java:
```diff
@@ -0,0 +1,56 @@
+           @RequestParam(name = "from", defaultValue = "0") @PositiveOrZero Long from,
+           @RequestParam(name = "size", defaultValue = "10") @Positive int size) {
+
+      return new ResponseEntity<>(service.getUsers(ids, from, size), HttpStatus.OK);
```
Объект Pageable можно создавать сразу здесь, чтобы не тянуть параметры from & size через несколько слоев

Kirill495 replied:
Исправил

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/category/mapper/CategoryMapper.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.List;
+
+@Mapper(componentModel = MappingConstants.ComponentModel.SPRING,
```
Супер, что используете MapStruct

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/category/storage/entity/CategoryEntity.java:
```diff
@@ -0,0 +1,23 @@
+
+@Entity
+@Table(name = "categories")
+@Data
```
Для диплома требование обязательное 

В аннотации Data есть реализация методов equals & hashCode и это очень важные методы для сущности БД, поэтому не стоит отдавать их реализацию на откуп сторонней библиотеки ломбок, а лучше сделать их самостоятельно. В целом  достаточно вместо data ставить Getter & Setter

Подробнее здесь:
https://thorben-janssen.com/lombok-hibernate-how-to-avoid-common-pitfalls/

ключевое ниже

Your equals() method should return false if the primary key value of both entity objects is null. But Lombok’s equals() method returns true. Because of that, you can’t add two new entity objects to a Set

Пример: 

```java
@Override
    public void run(String... args) throws Exception {
        Person person1 = new Person();
        Person person2 = new Person();

        System.out.println(person1.equals(person2)); // true
    }

    @Data
    class Person {
        private String name;
    }
```

Kirill495 replied:
Исправил

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/category/storage/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+   @Override
+   public List<CategoryDto> getCategories(int from, int size) {
+      Pageable page = PageRequest.of(from / size, size);
+      return mapper.toDto(repository.findAll(page).toList());
```
Здорово, что используете метод toList(), а не Collectors(collect.toList()) 
Первый производительнее

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/category/storage/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+
+   @Override
+   public void removeCategory(Long id) {
+      repository.findById(id).orElseThrow(() -> new CategoryNotFoundException(id));
```
Здесь идет обращение за сущностью, чтобы проверить что сущность есть, достаточно воспользоваться методом exists() из JPA
А то findById() тянет все связи

Kirill495 replied:
Исправил

Commit:
Спринт 18.
Дипломный проект. Разработка основного сервиса. Реорганизация пакетов и классов в модуле main

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/compilation/service/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+
+      Pageable pageable = PageRequest.of(filter.getFrom() / filter.getSize(), filter.getSize());
+      List<Compilation> compilations;
+      if (pinned) {
```
Можно использовать 1 запрос, просто передавать туда nullable фильтры. 

https://www.baeldung.com/spring-data-jpa-null-parameters#Ignoring

То есть принимать объект обертку Boolean

Kirill495 replied:
Исправил

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/compilation/service/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+      Pageable pageable = PageRequest.of(filter.getFrom() / filter.getSize(), filter.getSize());
+      List<Compilation> compilations = mapper.toModel(
+              repository.findAll(CompilationSpecification.isPinned(pinned), pageable).getContent());
+//      if (pinned) {
```
Закомментированный код нужно удалить

Kirill495 replied:
Исправил

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/event/storage/entity/EventEntity.java:
```diff
@@ -0,0 +1,76 @@
+import javax.persistence.Table;
+import java.time.LocalDateTime;
+
+@Data
```
Убрать Data

Kirill495 replied:
Исправил

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/event/storage/entity/EventEntity.java:
```diff
@@ -0,0 +1,76 @@
+   @Column(name = "annotation")
+   private String annotation;
+
+   @ManyToOne
```
У ManyToOne нужно переопределять fetchType = LAZY, т.к. по умолчанию стоит EAGER 
И при любом чтении сущности - подгружается всегда сторона отношения

VladimirTitov4 replied:
пройтись везде

Kirill495 replied:
Исправил

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/participation/storage/entity/RequestEntity.java:
```diff
@@ -0,0 +1,46 @@
+import javax.persistence.Table;
+import java.time.LocalDateTime;
+
+@Data
```
Data & Lazy

Kirill495 replied:
Исправил

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/user/storage/specification/UserSpecification.java:
```diff
@@ -0,0 +1,19 @@
+
+   @Override
+   public Predicate toPredicate(Root<UserEntity> root, CriteriaQuery<?> query, CriteriaBuilder criteriaBuilder) {
+      return null;
```
Здесь точно все ок ?

Kirill495 replied:
Удалил.

Commit:
Спринт 18.
Дипломный проект. Разработка основного сервиса. Доработка по замечаниям ревьюера

Commit:
Спринт 18.
Исправил ошибку в удалении категории

VladimirTitov4 commented on main-service/src/main/java/ru/practicum/ewm/main_service/compilation/service/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,113 @@
+
+   private void appendStatsToCompilations(List<Compilation> compilations) {
+      Set<Event> events = compilations.stream()
+              .flatMap(compilation -> (compilation.getEvents()).stream())
```
(compilation.getEvents()) не обязательно в скобки оборачивать
